Sources/SwiftCompilerPlugin/PluginMacroExpansionContext.swift

```diff
@@ -0,0 +1,253 @@
+/// used to get line/column from a syntax node in the managed source code.
+class SourceManager {
+  class KnownSourceSyntax {
+    struct Location {
+      var offset: Int
+      var line: Int
+      var column: Int
+      var fileID: String
+      var fileName: String
+    }
+
+    let node: Syntax
+    let location: Location
+
+    init(node: Syntax, location: Location) {
+      self.node = node
+      self.location = location
+    }
+
+    /// Location converter to get line/column in the node.
+    lazy var locationConverter: SourceLocationConverter = .init(
+      file: self.location.fileName,
+      tree: self.node
+    )
+  }
+
+  struct SourcePosition {
+    var fileName: String
+    var utf8Offset: Int
+  }
+
+  struct SourceRange {
+    var fileName: String
+    var startUTF8Offset: Int
+    var endUTF8Offset: Int
+  }
+
+  /// Syntax added by `add(_:)` method. Keyed by the `id` of the node.
+  private var knownSourceSyntax: [Syntax.ID: KnownSourceSyntax] = [:]
+
+  /// Convert syntax information to a `Syntax` node. The location informations
+  /// are cached in the source manager to provide `location(of:)` et al.
+  func add(_ syntaxInfo: PluginMessage.Syntax) -> Syntax {
+
+    var node: Syntax
+    var parser = Parser(syntaxInfo.source)
+    switch syntaxInfo.kind {
+    case .declaration:
+      node = Syntax(DeclSyntax.parse(from: &parser))
+    case .statement:
+      node = Syntax(StmtSyntax.parse(from: &parser))
+    case .expression:
+      node = Syntax(ExprSyntax.parse(from: &parser))
+    case .type:
+      node = Syntax(TypeSyntax.parse(from: &parser))
+    case .pattern:
+      node = Syntax(PatternSyntax.parse(from: &parser))
+    case .attribute:
+      node = Syntax(AttributeSyntax.parse(from: &parser))
+    }
+    do {
+      node = try OperatorTable.standardOperators.foldAll(node)
+    } catch let error {
+      // TODO: Error handling.
+      fatalError(error.localizedDescription)
+    }
+
+    let location = KnownSourceSyntax.Location(
+      offset: syntaxInfo.location.offset,
+      line: syntaxInfo.location.line,
+      column: syntaxInfo.location.column,
+      fileID: syntaxInfo.location.fileID,
+      fileName: syntaxInfo.location.fileName
+    )
+
+    knownSourceSyntax[node.id] = KnownSourceSyntax(node: node, location: location)
+
+    return node
+  }
+
+  /// Get position (file name + offset) of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func position(
+    of node: Syntax,
+    at kind: PositionInSyntaxNode
+  ) -> SourcePosition? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let localPosition = node.position(at: kind)
+    let positionOffset = base.location.offset
+    return SourcePosition(
+      fileName: base.location.fileName,
+      utf8Offset: localPosition.advanced(by: positionOffset).utf8Offset
+    )
+  }
+
+  /// Get position range of `node` in the known root nodes.
```

Comment:
Did you mean

```suggestion
  /// Get range of `node` in the known root nodes.
```

Reply:
Update the comment to clarify this is a range of "position" (filename + UTF-8 offset) not range of "location"  (file, line, col, and offset)

Reply:
Ah, makes sense. Thanks